entropy/coin_flip/app/src/flip_coin.ts

```diff
@@ -29,75 +46,82 @@ const parser = yargs(hideBin(process.argv))
+  console.log("1. Generating user's random number...");
 
-  const provider = new HDWalletProvider({
-    privateKeys: [privateKey],
-    providerOrUrl: rpc,
-  });
+  const randomNumber: `0x${string}` = `0x${crypto
+    .randomBytes(32)
+    .toString("hex")}`;
+  console.log(`User Generated Random number: ${randomNumber}`);
 
-  const web3 = new Web3(provider as any);
+  console.log("\n2. Requesting coin flip...");
 
-  const coinFlipContract = new web3.eth.Contract(
-    CoinFlipAbi as any,
-    coinFlipContractAddress
-  );
+  const flipFee = await coinFlipContract.read.getFlipFee();
+  console.log(`Flip Fee: ${flipFee} wei`);
 
-  console.log(`Running coin flip prototcol.`);
+  console.log("\n3. Sending request to flip coin...");
 
-  console.log("1. Generating user's random number...");
-  const randomNumber = web3.utils.randomHex(32);
-  console.log(`   number    : ${randomNumber}`);
-
-  console.log("2. Requesting coin flip...");
-  const flipFee = await coinFlipContract.methods.getFlipFee().call();
-  console.log(`   fee       : ${flipFee} wei`);
-
-  const receipt = await coinFlipContract.methods
-    .requestFlip(randomNumber)
-    .send({ value: flipFee, from: provider.getAddress(0) });
-
-  console.log(`   tx        : ${receipt.transactionHash}`);
-  const sequenceNumber = receipt.events.FlipRequest.returnValues.sequenceNumber;
-  console.log(`   sequence  : ${sequenceNumber}`);
-
-  console.log("3. Waiting for result...");
-  // Poll for new FlipResult events emitted by the CoinFlip contract. It checks if the event
-  // has the same sequenceNumber as the request. If it does,
-  // it logs the result and stops polling.
-  let fromBlock = receipt.blockNumber;
-  const intervalId = setInterval(async () => {
-    const currentBlock = await web3.eth.getBlockNumber();
-
-    if (fromBlock > currentBlock) {
-      return;
-    }
-
-    // Get 'FlipResult' events emitted by the CoinFlip contract for given block range.
-    const events = await coinFlipContract.getPastEvents("FlipResult", {
-      fromBlock: fromBlock,
-      toBlock: currentBlock,
+  const flipTxHash = await coinFlipContract.write.requestFlip([randomNumber], {
+    value: flipFee,
+  });
+  console.log(`Transaction Hash: ${flipTxHash}`);
+
+  const receipt = await client.waitForTransactionReceipt({
+    hash: flipTxHash,
+  });
+
+  const logs = parseEventLogs({
+    abi: ICoinFlipAbi,
+    eventName: "FlipRequest",
+    logs: receipt.logs,
+  });
+
+  const sequenceNumber = logs[0].args.sequenceNumber;
+
+  console.log(`\nSequence Number: ${sequenceNumber}`);
+
+  console.log("\n4. Waiting for flip result...");
+  const result = await new Promise((resolve, reject) => {
+  const unwatch = coinFlipContract.watchEvent.FlipResult({
+    fromBlock: receipt.blockNumber - 1n,
+    onLogs: (logs) => {
+        for (const log of logs) {
+          if (log.args.sequenceNumber === sequenceNumber) {
+            resolve(log.args.isHeads ? "Heads" : "Tails");
+            unwatch();
```

Comment:
control-flow wise it's probably better to unwatch before resolving -- it _shouldn't_ matter in most runtimes but I suppose it's possible that the runtime will return execution flow to the code awaiting the promise immediately upon `resolve`, meaning that `unwatch` wouldn't get called until the event loop flushes.  That wouldn't be "correct" behavior of promises but it is, I suppose, possible, so it's probably best to have the example code clean up the watch _before_ resolving the promise.